File: servlets/src/java/edu/rice/rubis/servlets/SearchItemsByRegion.java
/*
 * Copyright (C) 2002-2009  OW2 Consortium
 *
 * This file is part of dcsj-rubis (below referred to as "this program").
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package edu.rice.rubis.servlets;

import java.io.IOException;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;

import javax.servlet.ServletException;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

/**
 * Build the html page with the list of all items for given category and region.
 *
 * @author <a href="mailto:cecchet@rice.edu">Emmanuel Cecchet</a>
 * @author <a href="mailto:julie.marguerite@inrialpes.fr">Julie Marguerite</a>
 * @author <a href="mailto:marco.guazzone@gmail.com">Marco Guazzone</a>
 */
public class SearchItemsByRegion extends RubisHttpServlet
{
	@Override
	protected void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException, ServletException
	{
		int categoryId, regionId;
		int page;
		int nbOfItems;

		ServletPrinter sp = null;
		sp = new ServletPrinter(response, "SearchItemsByRegion");

		String value = request.getParameter("category");
		if ((value == null) || (value.equals("")))
		{
			this.printError("You must provide a category!", sp);
			return;
		}
		categoryId = Integer.parseInt(value);

		value = request.getParameter("region");
		if ((value == null) || (value.equals("")))
		{
			this.printError("You must provide a region!", sp);
			return;
		}
		regionId = Integer.parseInt(value);

		value = request.getParameter("page");
		if ((value == null) || (value.equals("")))
		{
			page = 0;
		}
		else
		{
			page = Integer.parseInt(value);
		}

		value = request.getParameter("nbOfItems");
		if ((value == null) || (value.equals("")))
		{
			nbOfItems = 25;
		}
		else
		{
			nbOfItems = Integer.parseInt(value);
		}

		sp.printHTMLheader("RUBiS: Search items by region");
		this.itemList(categoryId, regionId, page, nbOfItems, sp);
		sp.printHTMLfooter();
	}

	@Override
	protected void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException, ServletException
	{
		this.doGet(request, response);
	}

	@Override
	protected int getPoolSize()
	{
		return Config.SearchItemsByRegionPoolSize;
	}

	/**
	 * Close both statement and connection.
	 */
	private void closeConnection(PreparedStatement stmt, Connection conn)
	{
		try
		{
			if (stmt != null)
			{
				stmt.close(); // close statement
			}
			if (conn != null)
			{
				conn.setAutoCommit(true);
				this.releaseConnection(conn);
			}
		}
		catch (Exception ignore)
		{
		}
	}

	/**
	 * Display an error message.
	 * @param errorMsg the error message value
	 */
	private void printError(String errorMsg, ServletPrinter sp)
	{
		this.printError("Search Items by Region", errorMsg, sp);
	}

	/** List items in the given category for the given region */
	private void itemList(int categoryId, int regionId, int page, int nbOfItems, ServletPrinter sp)
	{
		String itemName, endDate;
		int itemId, nbOfBids = 0;
		float maxBid;
		ResultSet rs = null;
		PreparedStatement stmt = null;
		Connection conn = null;

		// get the list of items
		try
		{
			conn = this.getConnection();
			stmt = conn.prepareStatement("SELECT items.name, items.id, items.end_date, items.max_bid, items.nb_of_bids, items.initial_price FROM items,users WHERE items.category=? AND items.seller=users.id AND users.region=? AND end_date>=NOW() ORDER BY items.end_date ASC LIMIT ?,?");
			stmt.setInt(1, categoryId);
			stmt.setInt(2, regionId);
			stmt.setInt(3, page * nbOfItems);
			stmt.setInt(4, nbOfItems);
			rs = stmt.executeQuery();
		}
		catch (Exception e)
		{
			this.printError("Failed to execute Query for items in region: " + e, sp);
			this.closeConnection(stmt, conn);
			return;
		}
		try
		{
			if (!rs.first())
			{
				if (page == 0)
				{
					sp.printHTML("<h3>Sorry, but there is no items in this category for this region.</h3><br>");
				}
				else
				{
					sp.printHTML("<h3>Sorry, but there is no more items in this category for this region.</h3><br>");
					sp.printItemHeader();
					sp.printItemFooter("<a href=\"/rubis_servlets/servlet/edu.rice.rubis.servlets.SearchItemsByRegion?category="
									+ categoryId
									+ "&region="
									+ regionId
									+ "&page="
									+ (page - 1)
									+ "&nbOfItems="
									+ nbOfItems
									+ "\">Previous page</a>",
									"");
				}
				this.closeConnection(stmt, conn);
				return;
			}

			sp.printItemHeader();
			do
			{
				itemName = rs.getString("name");
				itemId = rs.getInt("id");
				endDate = rs.getString("end_date");
				maxBid = rs.getFloat("max_bid");
				nbOfBids = rs.getInt("nb_of_bids");
				float initialPrice = rs.getFloat("initial_price");
				if (maxBid < initialPrice)
				{
					maxBid = initialPrice;
				}
				sp.printItem(itemName, itemId, maxBid, nbOfBids, endDate);
			}
			while (rs.next());
			if (page == 0)
			{
				sp.printItemFooter("",
								"<a href=\"/rubis_servlets/servlet/edu.rice.rubis.servlets.SearchItemsByRegion?category="
								+ categoryId
								+ "&region="
								+ regionId
								+ "&page="
								+ (page + 1)
								+ "&nbOfItems="
								+ nbOfItems
								+ "\">Next page</a>");
			}
			else
			{
				sp.printItemFooter("<a href=\"/rubis_servlets/servlet/edu.rice.rubis.servlets.SearchItemsByRegion?category="
								+ categoryId
								+ "&region="
								+ regionId
								+ "&page="
								+ (page - 1)
								+ "&nbOfItems="
								+ nbOfItems
								+ "\">Previous page</a>",
								"<a href=\"/rubis_servlets/servlet/edu.rice.rubis.servlets.SearchItemsByRegion?category="
								+ categoryId
								+ "&region="
								+ regionId
								+ "&page="
								+ (page + 1)
								+ "&nbOfItems="
								+ nbOfItems
								+ "\">Next page</a>");
			}
		}
		catch (Exception e)
		{
			this.printError("Exception getting item list: " + e, sp);
		}
		this.closeConnection(stmt, conn);
	}
}

File: php/SearchItemsByRegion.php
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html>
  <body>
    <?php
    $scriptName = "SearchItemsByCategories.php";
    include("PHPprinter.php");
    $startTime = getMicroTime();
    
	$regionId = NULL;
	if (isset($_POST['region']))
	{
    	$regionId = $_POST['region'];
	}
	else if (isset($_GET['region']))
	{
    	$regionId = $_GET['region'];
	}
	else
	{
		printError($scriptName, $startTime, "Search Items By Region", "You must provide a region!<br>");
		exit();
	}
	$categoryId = NULL;
	if (isset($_POST['category']))
	{
    	$categoryId = $_POST['category'];
	}
	else if (isset($_GET['category']))
	{
    	$categoryId = $_GET['category'];
	}
	else
	{
		printError($scriptName, $startTime, "Search Items By Region", "You must provide a category identifier!<br>");
		exit();
	}
	$categoryName = NULL;
	if (isset($_POST['categoryName']))
	{
    	$categoryName = $_POST['categoryName'];
	}
	else if (isset($_GET['categoryName']))
	{
    	$categoryName = $_GET['categoryName'];
	}
	else
	{
		printError($scriptName, $startTime, "Search Items By Region", "You must provide a category name!<br>");
		exit();
	}
	$page = NULL;
	if (isset($_POST['page']))
	{
    	$page = $_POST['page'];
	}
	else if (isset($_GET['page']))
	{
    	$page = $_GET['page'];
	}
	else
	{
		$page = 0;
	}
	$nbOfItems = NULL;
	if (isset($_POST['nbOfItems']))
	{
    	$nbOfItems = $_POST['nbOfItems'];
	}
	else if (isset($_GET['nbOfItems']))
	{
    	$nbOfItems = $_GET['nbOfItems'];
	}
	else
	{
		$nbOfItems = 25;
	}

    printHTMLheader("RUBiS: Search items by region");
    print("<h2>Items in category $categoryName</h2><br><br>");
    
    getDatabaseLink($link);
    $result = mysql_query("SELECT items.id,items.name,items.initial_price,items.max_bid,items.nb_of_bids,items.end_date FROM items,users WHERE items.category=$categoryId AND items.seller=users.id AND users.region=$regionId AND end_date>=NOW() LIMIT ".$page*$nbOfItems.",$nbOfItems");
	if (!$result)
	{
		error_log("[".__FILE__."] Query 'SELECT items.id,items.name,items.initial_price,items.max_bid,items.nb_of_bids,items.end_date FROM items,users WHERE items.category=$categoryId AND items.seller=users.id AND users.region=$regionId AND end_date>=NOW() LIMIT ".$page*$nbOfItems."' failed: " . mysql_error($link));
		die("ERROR: Query failed for category '$categoryId', region '$regionId', page '$page' and nbOfItems '$nbOfItems': " . mysql_error($link));
	}
    if (mysql_num_rows($result) == 0)
    {
      if ($page == 0)
        print("<h3>Sorry, but there is no item in this category for this region.</h3><br>\n");
      else
      {
        print("<h2>Sorry, but there are no more items available in this category for this region!</h2>");
        print("<p><CENTER>\n<a href=\"/PHP/SearchItemsByRegion.php?category=$categoryId&region=$regionId".
              "&categoryName=".urlencode($categoryName)."&page=".($page-1)."&nbOfItems=$nbOfItems\">Previous page</a>\n</CENTER>\n");
      }
      mysql_free_result($result);
      mysql_close($link);
      printHTMLfooter($scriptName, $startTime);
      exit();
    }
    else
      print("<TABLE border=\"1\" summary=\"List of items\">".
            "<THEAD>".
            "<TR><TH>Designation<TH>Price<TH>Bids<TH>End Date<TH>Bid Now".
            "<TBODY>");

    while ($row = mysql_fetch_array($result))
    {
      $maxBid = $row["max_bid"];
      if ((is_null($maxBid)) ||($maxBid == 0))
	$maxBid = $row["initial_price"];

      print("<TR><TD><a href=\"/PHP/ViewItem.php?itemId=".$row["id"]."\">".$row["name"].
            "<TD>$maxBid".
            "<TD>".$row["nb_of_bids"].
            "<TD>".$row["end_date"].
            "<TD><a href=\"/PHP/PutBidAuth.php?itemId=".$row["id"]."\"><IMG SRC=\"/PHP/bid_now.jpg\" height=22 width=90></a>");
    }
    print("</TABLE>");  
    if ($page == 0)
      print("<p><CENTER>\n<a href=\"/PHP/SearchItemsByRegion.php?category=$categoryId&region=$regionId".
           "&categoryName=".urlencode($categoryName)."&page=".($page+1)."&nbOfItems=$nbOfItems\">Next page</a>\n</CENTER>\n");
    else
      print("<p><CENTER>\n<a href=\"/PHP/SearchItemsByRegion.php?category=$categoryId&region=$regionId".
            "&categoryName=".urlencode($categoryName)."&page=".($page-1)."&nbOfItems=$nbOfItems\">Previous page</a>\n&nbsp&nbsp&nbsp".
            "<a href=\"/PHP/SearchItemsByRegion.php?category=$categoryId&region=$regionId".
            "&categoryName=".urlencode($categoryName)."&page=".($page+1)."&nbOfItems=$nbOfItems\">Next page</a>\n\n</CENTER>\n");

    mysql_free_result($result);
    mysql_close($link);
    
    printHTMLfooter($scriptName, $startTime);
    ?>
  </body>
</html>
